File: libs/go/alpflags/cache.go
package alpflags

import (
	"sync"
	"time"
)

type cacheKey struct {
	flag   string
	tenant string // empty string == no tenant
}

type cacheEntry struct {
	value     bool
	expiresAt time.Time
}

// ttlCache is a thread-safe TTL cache keyed by (flag, tenant).
// invalidate(flag) drops every tenant entry for that flag in one shot.
type ttlCache struct {
	mu  sync.Mutex
	ttl time.Duration
	m   map[cacheKey]cacheEntry
}

func newTTLCache(ttl time.Duration) *ttlCache {
	return &ttlCache{ttl: ttl, m: make(map[cacheKey]cacheEntry)}
}

func (c *ttlCache) get(k cacheKey) (bool, bool) {
	c.mu.Lock()
	defer c.mu.Unlock()
	e, ok := c.m[k]
	if !ok {
		return false, false
	}
	if time.Now().After(e.expiresAt) {
		delete(c.m, k)
		return false, false
	}
	return e.value, true
}

func (c *ttlCache) put(k cacheKey, v bool) {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.m[k] = cacheEntry{value: v, expiresAt: time.Now().Add(c.ttl)}
}

// invalidate drops every entry for the given flag (across all tenants).
func (c *ttlCache) invalidate(flag string) {
	c.mu.Lock()
	defer c.mu.Unlock()
	for k := range c.m {
		if k.flag == flag {
			delete(c.m, k)
		}
	}
}

func (c *ttlCache) size() int {
	c.mu.Lock()
	defer c.mu.Unlock()
	return len(c.m)
}

File: libs/python/alp_flags/src/alp_flags/cache.py
"""In-memory TTL cache keyed by (flag_name, tenant_id_or_None)."""

from __future__ import annotations

import time
from threading import Lock
from typing import Any

CacheKey = tuple[str, str | None]


class TtlCache:
    def __init__(self, ttl_seconds: float) -> None:
        self._ttl = ttl_seconds
        self._data: dict[CacheKey, tuple[Any, float]] = {}
        self._lock = Lock()

    def get(self, key: CacheKey) -> Any | None:
        with self._lock:
            entry = self._data.get(key)
            if entry is None:
                return None
            value, expires_at = entry
            if time.monotonic() >= expires_at:
                self._data.pop(key, None)
                return None
            return value

    def put(self, key: CacheKey, value: Any) -> None:
        with self._lock:
            self._data[key] = (value, time.monotonic() + self._ttl)

    def invalidate(self, flag_name: str) -> None:
        """Drop all entries (across tenants) for a single flag."""
        with self._lock:
            for key in [k for k in self._data if k[0] == flag_name]:
                self._data.pop(key, None)

    def clear(self) -> None:
        with self._lock:
            self._data.clear()
